### PR TITLE
ENH: Make swapaxes always return a view. Fixes #5260

### LINKED ISSUE #5260: np.swapaxes doesn't always return a view
Related somehow to the discussion in #5258. The docs for `np.swapaxes` state that it returns a view if the first argument is an ndarray, but it has a fast path for `axis1 == axis2`, where it actually returns the original array incref'ed:

```
>>> a = np.arange(60).reshape(3, 4, 5).copy()
>>> a.base is None
True
>>> a.swapaxes(0, 1).base is None
False
>>> a.swapaxes(0, 0).base is None
True
>>> a.swapaxes(20, 20).base is None  # that this works is bad!!!
True
>>> a.transpose(range(3)).base is None
False
```

I prefer that it always returned a view, even if it is of the whole array unchanged, as the docs say, and as e.g. `np.transpose` does, even if the array is unchanged, and will be more than happy to submit a PR. If there is any good reason why this should not be the case, then at least the docs should be updated, and the check for `axis1 == axis2` be moved to after they have both been checked for in-boundness.

### PR DESCRIPTION
**Pull request by jaimefrio**

### CODE REVIEW

File: numpy/core/src/multiarray/shape.c [hunk 1]
```diff
@@ -683,25 +672,42 @@ PyArray_SwapAxes(PyArrayObject *ap, int a1, int a2)
...
-            val = a2;
-        }
-        else if (i == a2) {
-            val = a1;
+    if (a1 == a2) {
+        /* Return a view of the original array */
+        PyObject *ret;
+
+        Py_INCREF(PyArray_DESCR(ap));
+        ret = PyArray_NewFromDescr(Py_TYPE(ap), PyArray_DESCR(ap), n,
```
Comment (charris) [inline numpy/core/src/multiarray/shape.c]: Any idea what the last argument does? It isn't documented... OK, looks like it is needed for `__finalize__`, can be `NULL`.

Comment (jaimefrio) [inline numpy/core/src/multiarray/shape.c]: I used the code in `PyArray_Transpose`, which is the next function after this, as a template. We could actually eliminate this fast path altogether and let `PyArray_Transpose` handle the degenerate case also. There will be some unnecessary overhead, but perhaps simplicity is a better tradeoff than performance here.

**Comment by charris** [comment]
Hmm, looks like `PyArray_Transpose` always returns a view and will work if `a1 == a2`, so all that is needed is to avoid returning the original array directly. I wouldn't worry about a fast paths here.

This PR affects the C-API, so that should be documented. I'm inclined to think of this as a buglet fix, as the documentation says it returns a view. It's hard to think of a case where it could cause problems that wasn't due to a user bug.

**Comment by charris** [comment]
Apart from that, LGTM ;)

**Comment by jaimefrio** [comment]
Removed the fast path and added a *C API* remark to the release notes.

**Comment by charris** [comment]
Merged, thanks. BTW, I grepped for `PyArray_SetBaseObject` and found 35 spots that may be views being created. Could be worth looking at making a function for that. That would make it easy to add a fast path if we wanted to.

**Comment by jaimefrio** [comment]
That would be a new C API function, right?

**Comment by charris** [comment]
Could be. It would be interesting to see how common it is in extension code.

**Pull request merged**